Mathlib/Combinatorics/SetFamily/FourFunctions.lean

```diff
@@ -0,0 +1,356 @@
+* `Finset.le_card_infs_mul_card_sups`: Daykin inequality. `|s| |t| ≤ |s ⊼ t| |s ⊻ t|`
+* `holley`: Holley inequality.
+* `fkg`: Fortuin-Kastelyn-Ginibre inequality.
+* `Finset.card_le_card_diffs`: Marica-Schönheim inequality. `|s| ≤ |{a \ b | a, b ∈ s}|`
+
+## TODO
+
+Prove that lattices in which `Finset.le_card_infs_mul_card_sups` holds are distributive. See
+Daykin, *A lattice is distributive iff |A| |B| <= |A ∨ B| |A ∧ B|*
+
+Prove the Fishburn-Shepp inequality.
+
+Is `collapse` a construct generally useful for set family inductions? If so, we should move it to an
+earlier file and give it a proper API.
+-/
+
+open Finset Fintype
+open scoped BigOperators FinsetFamily
+
+variable {α β : Type*}
+
+section Finset
+variable [DecidableEq α] [LinearOrderedCommSemiring β] [ExistsAddOfLE β] {𝒜 ℬ : Finset (Finset α)}
+  {a : α} {f f₁ f₂ f₃ f₄ g μ : Finset α → β} {s t u : Finset α}
+
+/-- The `n = 1` case of the Ahlswede-Daykin inequality. Note that we can't just expand everything
+out and bound termwise since `c₀ * d₁` appears twice on the RHS of the assumptions while `c₁ * d₀`
+does not appear. -/
+private lemma ineq {a₀ a₁ b₀ b₁ c₀ c₁ d₀ d₁ : β}
+  (ha₀ : 0 ≤ a₀) (ha₁ : 0 ≤ a₁) (hb₀ : 0 ≤ b₀) (hb₁ : 0 ≤ b₁)
+  (hc₀ : 0 ≤ c₀) (hc₁ : 0 ≤ c₁) (hd₀ : 0 ≤ d₀) (hd₁ : 0 ≤ d₁)
+  (h₀₀ : a₀ * b₀ ≤ c₀ * d₀) (h₁₀ : a₁ * b₀ ≤ c₀ * d₁)
+  (h₀₁ : a₀ * b₁ ≤ c₀ * d₁) (h₁₁ : a₁ * b₁ ≤ c₁ * d₁) :
+    (a₀ + a₁) * (b₀ + b₁) ≤ (c₀ + c₁) * (d₀ + d₁) := by
+  calc
+    _ = a₀ * b₀ + (a₀ * b₁ + a₁ * b₀) + a₁ * b₁ := by ring
+    _ ≤ c₀ * d₀ + (c₀ * d₁ + c₁ * d₀) + c₁ * d₁ := add_le_add_three h₀₀ ?_ h₁₁
+    _ = (c₀ + c₁) * (d₀ + d₁) := by ring
+  obtain hcd | hcd := (mul_nonneg hc₀ hd₁).eq_or_gt
+  · rw [hcd] at h₀₁ h₁₀
+    rw [h₀₁.antisymm, h₁₀.antisymm, add_zero] <;> positivity
+  refine' le_of_mul_le_mul_right _ hcd
+  calc (a₀ * b₁ + a₁ * b₀) * (c₀ * d₁)
+      = a₀ * b₁ * (c₀ * d₁) + c₀ * d₁ * (a₁ * b₀) := by ring
+    _ ≤ a₀ * b₁ * (a₁ * b₀) + c₀ * d₁ * (c₀ * d₁) := mul_add_mul_le_mul_add_mul h₀₁ h₁₀
+    _ = a₀ * b₀ * (a₁ * b₁) + c₀ * d₁ * (c₀ * d₁) := by ring
+    _ ≤ c₀ * d₀ * (c₁ * d₁) + c₀ * d₁ * (c₀ * d₁) :=
+        add_le_add_right (mul_le_mul h₀₀ h₁₁ (by positivity) $ by positivity) _
+    _ = (c₀ * d₁ + c₁ * d₀) * (c₀ * d₁) := by ring
+
+private def collapse (𝒜 : Finset (Finset α)) (a : α) (f : Finset α → β) (s : Finset α) : β :=
+  ∑ t in 𝒜.filter λ t ↦ t.erase a = s, f t
+
+private lemma erase_eq_iff (hs : a ∉ s) : t.erase a = s ↔ t = s ∨ t = insert a s := by
+  by_cases ht : a ∈ t <;>
+  · simp [ne_of_mem_of_not_mem', erase_eq_iff_eq_insert, *];
+    aesop
+
+private lemma filter_collapse_eq (ha : a ∉ s) (𝒜 : Finset (Finset α)) :
+    (𝒜.filter λ t ↦ t.erase a = s) =
+      if s ∈ 𝒜 then
+        (if insert a s ∈ 𝒜 then {s, insert a s} else {s})
+      else if
+        insert a s ∈ 𝒜 then {insert a s} else ∅ := by
+  ext t; split_ifs <;> simp [erase_eq_iff ha] <;> aesop
+
+lemma collapse_eq (ha : a ∉ s) (𝒜 : Finset (Finset α)) (f : Finset α → β) :
```

Comment:
I might be mistaken, but is this in the root namespace? I suppose everything up to L225 could be in a `FourFunctions` namespace?

Reply:
This is fine because `collapse` is `private`. Everything mentioning a `private` declaration in its statements is automatically `private` as well.